### PR TITLE
Be able to restart the RLS server

### LINKED ISSUE #129: Add command to restart the RLS
I can't find docs on how to do this exactly, but I expect one sends the `shutdown` and `exit` messages, then calls `stop`, then `start` on the language client. (But perhaps `stop` sends the LSP messages?).

### PR DESCRIPTION
**Pull request by genneth**
Tries to fix #129. Whilst this works, I'm not very happy with the structure. I think overall, we need a bit more scaffolding around separately storing things necessary for the extension, the workspace, etc. I've been looking at the source for vscode-powershell for inspiration.

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -29,41 +28,41 @@ function getSysroot(env: Object): string | Error {
     const rustcSysroot = child_process.spawnSync(
         'rustup', ['run', 'nightly', 'rustc', '--print', 'sysroot'], {env}
     );
-
+
```
Comment (nrc) [inline src/extension.ts]: Could you avoid inserting all this extra whitespace please

File: src/extension.ts [hunk 2]
```diff
@@ -29,41 +28,41 @@ function getSysroot(env: Object): string | Error {
...
+    
     if (!rustcSysroot.stdout || typeof rustcSysroot.stdout.toString !== 'function') {
         return new Error(`Couldn't get sysroot from \`rustc\`: Got no ouput`);
     }
-
+    
     const sysroot = rustcSysroot.stdout.toString()
-        .replace('\n', '').replace('\r', '');
-
+    .replace('\n', '').replace('\r', '');
```
Comment (nrc) [inline src/extension.ts]: or removing it :-)

File: src/extension.ts [hunk 3]
```diff
@@ -159,22 +167,32 @@ export function activate(context: ExtensionContext) {
...
-    const lc = new LanguageClient('Rust Language Server', serverOptions, clientOptions);
+    lc = new LanguageClient('Rust Language Server', serverOptions, clientOptions);
     lcOutputChannel = lc.outputChannel;
-
-    diagnosticCounter(lc);
-    registerCommands(lc, context);
-    activateTaskProvider();
-
+    
+    let runningDiagnostics = 0;
```
Comment (nrc) [inline src/extension.ts]: Why do you want to inline this code?

**Comment by genneth** [comment]
I tried to look for an unregister, but couldn't find it. I suspect it doesn't exist. The vscode-powershell essentially has global variables: it organises them carefully, inside an overall `SessionManager` object, the constructor of which does things like registering commands, and separately `start` and `stop` which can be called multiple times.

File: src/extension.ts (hunk 1 above)
Comment (genneth) [reply src/extension.ts]: Sure; a lot of this was over zealously letting VSCode decide on formatting.

File: src/extension.ts (hunk 3 above)
Comment (genneth) [reply src/extension.ts]: It didn't seem to be obviously a reusable function --- calling it repeatedly doesn't make sense?

**Comment by nrc** [comment]
I filed #152 for the refactoring

File: src/extension.ts (hunk 3 above)
Comment (nrc) [reply src/extension.ts]: It's more for self-documentation - I found it made the `activate` function easy to read by factoring out self-contained bits of functionality. Inlining them made for a long, hard to grok function.

**Comment by genneth** [comment]
Tweaked the patch to be minimal delta against master

**Comment by nrc** [comment]
Thank you!

**Pull request merged**